src/Cassandra/Requests/RequestExecution.cs

```diff
@@ -366,56 +370,94 @@ private void HandleException(Exception ex)
 
                 case RetryDecision.RetryDecisionType.Retry:
                     //Retry the Request using the new consistency level
-                    Retry(decision.RetryConsistencyLevel, decision.UseCurrentHost);
+                    Retry(retryInformation.Decision.RetryConsistencyLevel, retryInformation.Decision.UseCurrentHost, host);
                     break;
             }
-        }
 
+            _requestObserver.OnRequestError(host, retryInformation.Reason, retryInformation.Decision.DecisionType);
+        }
+        
         /// <summary>
-        /// Gets the retry decision based on the exception from Cassandra
+        /// Gets the retry decision based on the request error
         /// </summary>
-        internal static RetryDecision GetRetryDecision(
-            Exception ex, IExtendedRetryPolicy policy, IStatement statement, Configuration config, int retryCount)
+        internal static RetryDecisionWithReason GetRetryDecisionWithReason(
+            IRequestError error, IExtendedRetryPolicy policy, IStatement statement, Configuration config, int retryCount)
         {
-            if (ex is SocketException exception)
+            var ex = error.Exception;
+            if (ex is SocketException || ex is OverloadedException || ex is IsBootstrappingException || ex is TruncateException ||
+                ex is OperationTimedOutException)
             {
-                RequestExecution.Logger.Verbose("Socket error " + exception.SocketErrorCode);
-                return policy.OnRequestError(statement, config, exception, retryCount);
-            }
-            if (ex is OverloadedException || ex is IsBootstrappingException || ex is TruncateException)
-            {
-                return policy.OnRequestError(statement, config, ex, retryCount);
+                if (ex is SocketException exception)
+                {
+                    RequestExecution.Logger.Verbose("Socket error " + exception.SocketErrorCode);
+                }
+
+                // For PREPARE requests, retry on next host
```

Comment:
Just a note, in the future, we could avoid using `RequestHandler` / `RequestExecution` classes for `PREPARE` requests as the behaviour is so different from the rest. We could introduce something like a [`PrepareHandler`](https://github.com/datastax/nodejs-driver/blob/master/lib/prepare-handler.js) :)